CUDA support in the CSR layout: sparse_to_dense/add_sparse_csr

aocsa opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#56561 CUDA support in the CSR layout: sparse_to_dense/add_sparse_csr**
* #56560 CUDA support in the CSR layout: constructors
* #56744 Fix: Avoid unnecessary creation of tensors in SparseCsrTensorImpl::resize_and_clear_
* #56723 Fix: sparse_csr_tensor segfaults when crow_indices or col_indices are non-tensors
* #56722 Fix: Incorrect example output in sparse_csr_tensor doc-string

Closes gh-56485

Differential Revision: [D27939682](https://our.internmc.facebook.com/intern/diff/D27939682)

aocsa committed:
CUDA support in the CSR layout: sparse_to_dense/add_sparse_csr

[ghstack-poisoned]

aocsa committed:
Update on "CUDA support in the CSR layout: sparse_to_dense/add_sparse_csr"




[ghstack-poisoned]

aocsa committed:
Update on "CUDA support in the CSR layout: sparse_to_dense/add_sparse_csr"


Closes gh-56485

[ghstack-poisoned]

aocsa committed:
Update on "CUDA support in the CSR layout: sparse_to_dense/add_sparse_csr"


Closes gh-56485

[ghstack-poisoned]

aocsa committed:
Update on "CUDA support in the CSR layout: sparse_to_dense/add_sparse_csr"


Closes gh-56485

[ghstack-poisoned]

ezyang commented on aten/src/ATen/native/sparse/SparseCsrTensor.cpp:
```diff
@@ -50,15 +50,18 @@ Tensor sparse_csr_tensor(
+          crow_indices_accessor[0] == 0, "0th value of crow_indices must be 0.");
+    });
+  } else {
+    //@aocsa, TODO: cuda case
```
TODO

ezyang replied:
It seems like you're going to need a sync here. Is it possible some of this information should be stored on CPU, or perhaps maybe you should just do an async cuda assert here.

ezyang commented on aten/src/ATen/native/sparse/SparseCsrTensor.cpp:
```diff
@@ -95,18 +97,18 @@ Tensor sparse_csr_tensor(
-  std::array<int64_t, 2> size;
 
+  std::array<int64_t, 2> size;
+  auto host_col_indices = crow_indices.is_cpu() ? col_indices : col_indices.to(kCPU);
```
This looks a bit suspect. What's the invariant for when non-CPU is accepted or not?

ezyang commented on torch/_tensor.py:
```diff
@@ -926,17 +926,21 @@ def to_sparse_csr(self):
         if self.is_sparse:
             coalesced_self = self.coalesce()
-            row_indices = coalesced_self.indices()[0]
+            row_indices = coalesced_self.indices().select(0, 0)
```
not sure what's going on here

aocsa committed:
Update on "CUDA support in the CSR layout: sparse_to_dense/add_sparse_csr"


Closes gh-56485

Differential Revision: [D27939682](https://our.internmc.facebook.com/intern/diff/D27939682)

[ghstack-poisoned]

aocsa committed:
Update on "CUDA support in the CSR layout: sparse_to_dense/add_sparse_csr"


Closes gh-56485

Differential Revision: [D27939682](https://our.internmc.facebook.com/intern/diff/D27939682)

[ghstack-poisoned]

aocsa committed:
Update on "CUDA support in the CSR layout: sparse_to_dense/add_sparse_csr"


Closes gh-56485

Differential Revision: [D27939682](https://our.internmc.facebook.com/intern/diff/D27939682)

[ghstack-poisoned]

aocsa committed:
Update on "CUDA support in the CSR layout: sparse_to_dense/add_sparse_csr"


Closes gh-56485

Differential Revision: [D27939682](https://our.internmc.facebook.com/intern/diff/D27939682)

[ghstack-poisoned]

aocsa committed:
Update on "CUDA support in the CSR layout: sparse_to_dense/add_sparse_csr"


Closes gh-56485

Differential Revision: [D27939682](https://our.internmc.facebook.com/intern/diff/D27939682)

[ghstack-poisoned]

aocsa committed:
Update on "CUDA support in the CSR layout: sparse_to_dense/add_sparse_csr"


Closes gh-56485

Differential Revision: [D27939682](https://our.internmc.facebook.com/intern/diff/D27939682)

[ghstack-poisoned]

aocsa committed:
Update on "CUDA support in the CSR layout: sparse_to_dense/add_sparse_csr"


Closes gh-56485

Differential Revision: [D27939682](https://our.internmc.facebook.com/intern/diff/D27939682)

[ghstack-poisoned]
